pkg/trace/obfuscate/sql_tokenizer.go

```diff
@@ -528,37 +537,40 @@ func (tkn *SQLTokenizer) scanCommentType2() (TokenKind, []byte) {
 	return Comment, tkn.bytes()
 }
 
-// advance advances the tokenizer to the next rune. If this is not possible, tkn.lastChar will
-// be set to EOFChar.
+// advance advances the tokenizer to the next rune. If the decoder encounters an error decoding, or
+// the end of the buffer is reached, tkn.lastChar will be set to EOFChar. In case of a decoding
+// error, tkn.err will also be set.
 func (tkn *SQLTokenizer) advance() {
 	ch, n := utf8.DecodeRune(tkn.buf[tkn.off:])
-	if ch == utf8.RuneError && n == 0 {
-		// only EOF is possible
+	if ch == utf8.RuneError && (n == 0 || n == 1) {
 		tkn.pos++
 		tkn.lastChar = EOFChar
```

Comment:
This isn't true though when n == 1.

Reply:
This is actually an interesting problem. Bear with me while I explain my thinking.

I thought about using `LexError` but realized it is a `TokenKind` and not a `rune`.

We could define some other invalid rune value as a signal (say, `const EncodingError`) the same way we use `EOFChar` ([we can use the spec to determine what is valid](https://golang.org/ref/spec#Rune_literals)), but that doesn't get us very much since pretty much everywhere we would just change the condition `if tkn.lastChar == EOFChar` to `if tkn.lastChar == EOFChar || tkn.lastChar == EncodingError`. The one place having `EncodingError` *would* be useful is [here](https://github.com/DataDog/datadog-agent/pull/6727/files#diff-35f8e80fd41a02225875fde0050337f71a915bd7919d6bb0e6bd97b78ac34501R153) in this same PR, where we wouldn't have to check if `tkn.err != nil`, but it didn't strike me as important.

Except for that check, we don't really need to distinguish between encoding error and EOF. My reasoning for setting `lastChar = EOFChar` is that seeing an invalid encoding is **logically** an EOF since we don't continue to process the query.

I also considered fixing the encoding on the fly. The problem there is that the UTF-8 replacement character is three bytes, so replacing a single bad byte with the replacement character requires replacing one byte in the slice with three. This becomes not just runtime overhead, but also complicates the code for an uncommon edge case.

I'm open to defining `EncodingError` if you think that makes the code clearer. Let me know if there's another solution that I'm missing.

Reply:
I see. My bad. I didn't realise that one is a `TokenKind` and the other a `rune` 🤦 

I'm a bit concerned that `EOFChar` can induce confusion later, leading to believe people it's EOF instead of an actual decoding error. 

```go
// EOFChar is used to signal that no more characters were found by the scanner. It is
// an invalid rune value that can not be found in any valid string.
const EOFChar = unicode.MaxRune + 1
```
I have two ideas (not sure if any good):
1. Rename to `EndChar` and update docs
2. Introduce a new rune like you suggest, and have a new function `tkn.doneReading()` or similar which does the check to remove the duplication.
3. Keep as is and update docs.

Feel free to apply any and merge.

Reply:
P.S. You should probably never get invalid UTF-8 by the way, because it's fixed at decoding time.

Reply:
I'm partial to `#1` where we rename `EOFChar` to `EndChar`. We can already determine if there was an error by checking if `tkn.err` is non-nil. I would prefer to keep a single variable for signaling an error rather than having two.

Renaming to `EndChar` should be sufficient to de-confuse the situation I think.

And yes, good point about the encoding. Since the encoding is fixed during msgpack decoding, we *should* never encounter this issue - another reason why just quitting the query rather than trying to fix the encoding is ok.